Print exception and abort if no handler is found

Burgos opened the pull request:
Previously, dwarf exception handler was relying on a fact that running
user code is always enclosed in the catch-all exception handler.  That
might not be the case if the user (with debugger, for example) change
value of `rt_trapExceptions` to false. To cover this scenario, instead
of issuing fatal error, we wil just print exception info and abort the
program.

Burgos commented:
Example with `rt_trapException` cleared:

```
burgos@ubuntu-vbox:~$ dmd-dev -run test2.d
uncaught exception
dwarfeh(224) fatal error
Error: program killed by signal 6
burgos@ubuntu-vbox:~$  # with this patch
burgos@ubuntu-vbox:~$ dmd-dev -run test2.d
uncaught exception
object.Exception@test2.d(3): test exception
----------------
Error: program killed by signal 6
```

Burgos commented:
I'm not sure why coverage is complaining. I've added tests for this line to be hit. Could it be that coverage is ignoring this test?

wilzbach commented:
>I'm not sure why coverage is complaining. I've added tests for this line to be hit. Could it be that coverage is ignoring this test? 

The following command is run to check for coverage:

```
 TEST_COVERAGE="1" make -j$N -C . -f posix.mak MODEL=$MODEL unittest-debug
```
https://github.com/dlang/druntime/blob/master/circleci.sh#L81

adamdruppe commented:
i like it

nemanja-boric-sociomantic commented:
I'll do rebase (to fix Jenkins issue) and coverage enhancement later today.

Burgos commented:
Rebased and updated.

Only thing confuses me now is the coverage report:

https://codecov.io/gh/dlang/druntime/compare/f4095df4c07b0fefba76f5bd0b72972d901b9f86...8eefed3a9877d7029480cc82e4b5d2cf460d8b20/src/src/rt/dwarfeh.d#L225...227

You can see it reports these lines as not covered. But, they indeed are covered (see the test in the commit). Why this shows as not covered is because program writes out the coverage report at the end of the program. But in case of this test, program must abort, so it doesn't write these lines. I'm not sure if I can do anything about it?

Burgos commented:
Thanks @Calrama !

ZombineDev commented:
@schveiguy @jmdavis @MartinNowak can you please take a look? This PR should be straightforward to review.

ZombineDev commented:
@Burgos CodeCov says that it has not coverage data on `test/exceptions/src/rt_trap_exceptions.d` which probably explain why the lines you changed in `src/rt/dwarfeh.d` are not covered. Your changes are fine, the problem is most probably with the CodeCov setup.

nemanja-boric-sociomantic commented:
Thanks!
